ops/terraform/services/server/deploy.groovy

```diff
@@ -45,36 +50,179 @@ terraform apply \
+        // than 0 and then capture the first message
+        if (!messages.isEmpty()) {
+            msg = messages[0]
+
+            printServerRegressionMessage(msg)
+            sqs.deleteMessage(msg.receipt, sqsQueueUrl)
+
+            testRunResult = msg.body.result
+            return testRunResult == 'SUCCESS' ? true : false
+        } else {
+            println "[Attempt ${i + 1}/${maxRetries}] No messages posted to ${sqsQueueName} yet, waiting ${heartbeatInterval} seconds to check again..."
+        }
+
+        sleep(heartbeatInterval)
+    }
+}
+
+/* Checks for indications of a running server-regression lambda by looking for unconsumed SQS messages */
+boolean canServerRegressionRunProceed(String awsRegion, String sqsQueueName, String bfdEnv) {
+    println "Checking server-regression ${sqsQueueName} state..."
+
+    if (sqs.queueExists(sqsQueueName)) {
+        sqsQueueUrl = sqs.getQueueUrl(sqsQueueName)
+        println "Queue ${sqsQueueName} exists. Checking for messages in ${sqsQueueUrl}..."
+        serverRegressionSignalMsgs = sqs.receiveMessages(
+                sqsQueueUrl: sqsQueueUrl,
+                awsRegion: awsRegion,
+                maxMessages: 10,
+                visibilityTimeoutSeconds: 0,
+                waitTimeSeconds: 0)
+        if (serverRegressionSignalMsgs?.isEmpty()) {
+            println "Queue ${sqsQueueName} is empty. bfd-${bfdEnv}-server-regression run can proceed"
+            return true
+        } else {
+            println "Queue ${sqsQueueName} has messages. Is there an ongoing run of the server-regression lambda for the ${bfdEnv} environment?"
+            return false
+        }
+    } else {
+        println "Queue ${sqsQueueName} can not be found. Was the ${sqsQueueName} destroyed?"
+        return false
+    }
+}
+
+/* Print formatted messages about the outcome of the server-regression run */
+def printServerRegressionMessage(Map message) {
+    body = message.body
+
+    cloudWatchUrl = getCloudWatchLogUrl(
+        logGroupName: body.log_group_name,
+        logStreamName: body.log_stream_name
+    )
+    println "Timestamp: ${java.time.LocalDateTime.now().toString()}"
+    println """${body.function_name} finished running with a ${body.result} result, more information:
+                |   Lambda Request ID: ${body.request_id}
+                |   CloudWatch log stream name: ${body.log_stream_name}
+                |   CloudWatch log group name: ${body.log_group_name}
+                |   CloudWatch URL: ${cloudWatchUrl}""".stripMargin()
+}
+
+/* Gets the build ID of the last successful Jenkins job build of the current branch */
+def getLastSuccessfulBuildNum() {
+    def lastSuccessfulBuildID = 0
+    def build = currentBuild.previousBuild
+    while (build != null) {
+        if (build.result == "SUCCESS")
+        {
+            lastSuccessfulBuildID = build.id as Integer
+            break
+        }
+        build = build.previousBuild
```

Comment:
What happens if there is no previous build? Do we get 0? Will that case cause issues trying to compare to something it shouldnt?

Reply:
Yes, `0` will be returned for `lastSuccessfulBuildID`. However, that will not cause any problems as the stats comparison code will simply skip comparisons (and log that no applicable stats were found). Subsequent runs will work just fine.